src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+
+Quadratic forms over modules.
+-/
+
+import linear_algebra.bilinear_form
+import linear_algebra.determinant
+import linear_algebra.special_linear_group
+
+/-!
+# Quadratic forms
+
+This file defines quadratic forms over a `R`-module `M`.
+A quadratic form is a map `Q : M → R` such that
+  (`map_smul`) `Q (a • x) = a * a * Q x`
+  (`polar_...`) The map `polar Q := λ x y, Q (x + y) - Q x - Q y` is bilinear.
+They come with a scalar multiplication, `(a • Q) x = Q (a • x) = a * a * Q x`,
+and composition with linear maps `f`, `Q.comp f x = Q (f x)`.
+
+## Main definitions
+
+ * `quadratic_form.associated`: associated bilinear form
+ * `quadratic_form.pos_def`: positive definite quadratic forms
+ * `quadratic_form.discr`: discriminant of a quadratic form
+
+## Main statements
+
+ * `quadratic_form.associated_left_inverse`,
+ * `quadratic_form.associated_right_inverse`: in a commutative ring where 2 has
+  an inverse, there is a correspondence between quadratic forms and symmetric
+  bilinear forms
+
+## References
+
+ * https://en.wikipedia.org/wiki/Quadratic_form
+ * https://en.wikipedia.org/wiki/Discriminant#Quadratic_forms
+
+## Tags
+
+quadratic form, homogeneous polynomial, quadratic polynomial
+-/
+
+universes u v w
+variables {R : Type u} {M : Type v} [add_comm_group M] [ring R]
+variables {R₁ : Type u} [comm_ring R₁] [module R₁ M]
+
+/-- Up to a factor 2, `Q.polar` is the associated bilinear form for a quadratic form `Q`. -/
```

Comment:
I've never seen this terminology before. Out of curiosity, do you know where it comes from?

Reply:
I saw this name in [one section of the WIkipedia article](https://en.wikipedia.org/wiki/Quadratic_form#Generalization). I'm not aware of a name for it in other sources (because they go directly to `associated`). If you have a more common name, I'd be happy to change it.

Reply:
Oooh, no, I didn't have a name either. To me it was always a nameless gadget. So I'm fine with this name. But it might be good to add a pointer to wiki in the docstring, because the name will not ring a bell with many people.